Add OMotM and LCotM April

KrisXV opened the pull request:
@TheImmortal

Zarel commented on config/formats.js:
```diff
@@ -364,32 +364,49 @@ let Formats = [
-			battle: 2,
+		mod: 'linked',
+		ruleset: ['Gen 7] OU'],
+		banlist: ['King\'s Rock', 'Razor Fang'],
```
We use double quotes for names just so these don't need to be escaped.

KrisXV replied:
Since when? I've never seen that anywhere in `formats.js`

Zarel commented on config/formats.js:
```diff
@@ -364,32 +364,49 @@ let Formats = [
-			validate: [2, 4],
-			battle: 2,
+		mod: 'linked',
+		ruleset: ['Gen 7] OU'],
```
`[`

KrisXV replied:
Why do I always do this, I did it with STABmons too when it was first re-implemented

Marty-D commented on config/formats.js:
```diff
@@ -549,11 +567,28 @@ let Formats = [
+			validate: [2, 4],
+			battle: 2,
+		},
+		ruleset: ['Gen 7] Doubles OU'],
```
Also here, but this one was apparently around all last month.

KrisXV replied:
ok well thanks.

Slayer95 commented on mods/linked/scripts.js:
```diff
@@ -0,0 +1,504 @@
+	pokemon: {
+		moveUsed: function (move, targetLoc) {
+			let lastMove = this.moveThisTurn ? [this.moveThisTurn, this.battle.getMove(move).id] : this.battle.getMove(move).id;
+			this.lastMove = lastMove;
```
``pokemon.lastMove`` should be a ``Move`` instance now I think?

KrisXV replied:
yep

Slayer95 commented on mods/linked/statuses.js:
```diff
@@ -0,0 +1,78 @@
+	},
+
+	/**
+	 * Gems and Auras
```
This comment should be updated now that ``aura`` is no longer here.

KrisXV replied:
oops

Slayer95 commented:
Glad to see Linked back!

My main concerns over porting it to Gen 7 were Dancer and Instruct. I see Dancer is already handled somehow; however, Instruct is not mentioned anywhere. How does it behave? Is it specified in the Gen 7 Linked thread or up to implementation?

Slayer95 commented on mods/linked/moves.js:
```diff
@@ -0,0 +1,462 @@
+				this.add('-fail', source);
+				return null;
+			}
+			if (!target.linked && action.move.category === 'Status' && action.move.id !== 'mefirst') {
```
I don't think ``target.linked`` exists. For reference, this is from my XY implementation:

```js
	suckerpunch: {
		inherit: true,
		onTry: function (source, target) {
			let decision = this.willMove(target);
			if (!decision || decision.choice !== 'move') {
				this.add('-fail', source);
				return false;
			}
			if (target.volatiles['mustrecharge'] && target.volatiles['mustrecharge'].duration < 2) {
				// Duration may not be lower than 2 if Sucker Punch is used as a low-priority move
				// i.e. if Sucker Punch is linked with a negative priority move
				this.add('-fail', source);
				return false;
			}
			if (!decision.linked) {
				if (decision.move.category !== 'Status' || decision.move.id === 'mefirst') return;
				this.add('-fail', source);
				return false;
			}

			for (let i = 0; i < decision.linked.length; i++) {
				let linkedMove = this.getMove(decision.linked[i]);
				if (linkedMove.category !== 'Status' || linkedMove.id === 'mefirst') return;
			}
			this.add('-fail', source);
			return false;
		},
	},
```

So it should be ``action.linked`` now I guess?

KrisXV replied:
it was target.linked, idk why it got changed

Slayer95 replied:
Probably a bug then. Apply the patch pls :P
There is another bad instance of ``target.linked`` below too.

Slayer95 commented on mods/linked/moves.js:
```diff
@@ -0,0 +1,462 @@
+	},
+
+	/**
+	 * Copycat and Mirror Move
```
What happened to Copycat?

KrisXV replied:
removed because the code was unnecessary

Slayer95 commented on mods/linked/moves.js:
```diff
@@ -0,0 +1,462 @@
+			}
+		},
+	},
+};
```
Is Fusion Bolt + Fusion Flare no longer supported?
Or does it work correctly with default mechanics now?

KrisXV replied:
it works correctly

KrisXV replied:
just kidding, after testing, they dont work

KrisXV commented:
alright @Slayer95, i believe I fixed everything. I also spoke with that ivy guy and he just said to make Instruct work like other calling/copying moves, so I just adjusted it for ``getLastMoveAbsolute``. (also i fixed the stupid comma dangle thing)

VineST commented:
I don't remember ever getting specific with Instruct, but it makes sense to have it respect the last absolute move if there were to ever be a Doubles context.

Slayer95 commented on mods/linked/scripts.js:
```diff
@@ -0,0 +1,508 @@
+
+	pokemon: {
+		moveUsed: function (move, targetLoc) {
+			let lastMove = this.moveThisTurn ? [this.moveThisTurn, move] : move;
```
I don't think this actually affects gameplay, but ``this.moveThisTurn`` is a string, while ``move`` is a ``Move`` instance, so this is a weird tuple.

If you are interested, you might want to normalize the types.

Slayer95 commented:
Sorry for the late response. I just want to point out that this looks great, except for a tiny typing nit, noted in a comment above.
